test/ldap: assign non-busy ports to ldap

ptrsmrn opened the pull request:
It may happen that the ports we randomly choose for LDAP are busy, and that'd fail the test suite, so once we randomly select ports, now we'll see if they're busy or not, and if they're busy, we'll select next ones, until we finally have some free ports for LDAP.
Tested with: `./test.py ldap/ldap_connection_test --repeat 1000 -j 10`: before the fix, this command fails after 112 runs, and of course it passes with the fix.

Fixes: scylladb/scylla-enterprise#5120

No need to backport, it's just a flaky test fix.

nyh commented:
Also Fixes https://github.com/scylladb/scylladb/issues/23149 - right?

xtrey commented:
There is https://github.com/scylladb/scylladb/pull/23235 that reduces this issue, but not eliminate it completely. Also, it will be better to stick use IP from `HostRegistry()`, but this will require additional work for rewriting starting slapd.

nyh commented on test/pylib/suite/base.py:
```diff
@@ -399,14 +400,30 @@ def read_log(log_filename: pathlib.Path) -> str:
+def is_port_available(host, port):
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.settimeout(0.1)
+        return s.connect_ex((host, port)) != 0
```
It took me several minutes to understand how this function works because of the double-negative ;-) Basically, we try to connect to the given port number. If the connection succceds (in which case connect_ex returns 0) it means somebody is listening to this port so we can't, so we need to return false. In other words, we need to return True if connect_ex returns !=  0.

This is an interesting way to check for the availability of a port, why not just try to bind (not connect) to the port? Note that like in quantum mechanics, the mere act of connecting may do something ;-)

ptrsmrn replied:
When writing this, I was inspired by another method `def can_connect(address, family=socket.AF_INET)` from `prepare_instance.py`, which doesn't work for me - don't ask me why, IDK - but it also connects to the socket, not just binds to it. As for the value returned by `connect_ex`, can't do much about it, unless you have a better idea how to rewrite the return statement to make it less confusing.

nyh replied:
I see I stopped writing in mid-sentence - I tried to say at the end that for checking a port you connect to it, which may do something (imagine some other server there, and you connect to it). :-)
But never mind, I'm not against this function. I was just curious to understand how this function works and it took me a bit too much time to figure it out.

ptrsmrn commented on test/pylib/suite/base.py:
```diff
@@ -399,14 +400,30 @@ def read_log(log_filename: pathlib.Path) -> str:
         ldap_port = 5000 + (toxiproxy_id * 3) % 55000
+        while not can_connect_to_local_ports(range(ldap_port, ldap_port+3)):
+            print("Port {} is busy, trying next one".format(ldap_port))
+            ldap_port = ldap_port + 3
```
While you're technically correct, I can't imagine this would ever happen, so for the sake of the clarity of the code, I left just a "dumb" implementation. The situation now is that the test fails sporadically (cause we fail to allocate a free port), I don't expect we'll be forever retrying, because who would consume all 60k of ports - that'd be a bug in itself elsewhere, I suppose. Either way, I'll add a limit to the max port number we try in case something really goes south.

ptrsmrn replied:
Done.

ptrsmrn commented on test/pylib/suite/base.py:
```diff
@@ -399,14 +400,30 @@ def read_log(log_filename: pathlib.Path) -> str:
+        # Assign some NONE-busy ports to LDAP
         ldap_port = 5000 + (toxiproxy_id * 3) % 55000
+        while not can_connect_to_local_ports(range(ldap_port, ldap_port+3)):
+            print("Port {} is busy, trying next one".format(ldap_port))
```
Fixed, thanks.

ptrsmrn commented:
> > > Also Fixes #23149 - right?
> > 
> > 
> > Could be, this issue looks very much alike. There are other similar issues/duplicates, from what I remember. I have not linked all of them here, as I am not sure all of them will be fixed.
> 
> Then please make an effort to understand this, because the enterprise issue you are solving is 3 months old, but the issue #23149 happened multiple times in the last week and already has a bunch of duplicates. If you don't understand if these are duplicates that can be closed, nobody will :-(

OK, I think I found 2 other issues that look about the same, so this commit fixes 3 issues in total.

ptrsmrn commented:
> Proper way with `HostRegistry()` is ready #23235

@xtrey not sure I follow, should this PR use `HostRegistry` you introduce? I think this PR should get merged ASAP, at it kind of blocks the CI.

ptrsmrn commented:
UPDATE 1:
- addressed comments,
- attached 2 new issues, which this PR seems to solve (I'll monitor if they continue to appear after the PR is merged, but I'm positive it'll work).

xtrey commented:
> @xtrey not sure I follow, should this PR use HostRegistry you introduce? I think this PR should get merged ASAP, at it kind of blocks the CI.

No need to use `HostRegistry` there. I've implemented it in my refactoring for LDAP server that resolves issue with ports as well. This fix is not needed once toxiproxy lives on a separate host.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16001/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/16001/)
#### Build Details:
- Duration: 6 hr 37 min
- Builder: i-0a640198a9d59eab5 (m5d.16xlarge)

bhalevy commented:
So are we going with this fix or https://github.com/scylladb/scylladb/pull/23235?
In any case, it's killing CI and needs to be fixed ASAP.

nuivall commented on test/pylib/suite/base.py:
```diff
@@ -399,14 +400,33 @@ def read_log(log_filename: pathlib.Path) -> str:
+            print("One of {} ports required by LDAP are busy, trying the next 3 ports".format(list(range(ldap_port, ldap_port+3))))
+            ldap_port = ldap_port + 3
+            if ldap_port > 65535: # it's the highest possible TCP port number
+                print("No more ports to check, exiting")
```
if I understand the code that message is not true, it only means that `(ldap_port, ldap_port+3)` range ended, there are other ports to check

nuivall commented on test/pylib/suite/base.py:
```diff
@@ -399,14 +400,33 @@ def read_log(log_filename: pathlib.Path) -> str:
+    for port in ports_range:
+        if not is_port_available('localhost', port):
+            return False
+    return True
```
I think checking should somehow be based on exception from actually using the port, otherwise we can say the port is free but few lines later still fail to connect because someone else used it.

ptrsmrn replied:
I am using a function that doesn't throw, but am checking if we connect cleanly. Of course it can always happen that something connected in between (in between the setup and the code starting a server, but refactoring that is a bit bigger task).

ptrsmrn commented:
> So are we going with this fix or #23235? In any case, it's killing CI and needs to be fixed ASAP.

@bhalevy I think, whatever is first, since it's a pressing problem, and this is why I created this quick fix in the first place.

ptrsmrn commented on test/pylib/suite/base.py:
```diff
@@ -399,14 +400,33 @@ def read_log(log_filename: pathlib.Path) -> str:
         toxiproxy_id_gen += 1
+        # Assign some NONE-busy ports to LDAP
         ldap_port = 5000 + (toxiproxy_id * 3) % 55000
+        while not can_connect_to_local_ports(range(ldap_port, ldap_port+3)):
```
I didn't delve into the topic of why we start with these ports, just appended the logic of retry with the next ones on top of that. I think just retrying with next ones should be enough, if we run out of ports, it may mean something else is broken (and we can revisit this patch, this patch was meant to be simple, the problem is simple).

xemul commented:
Ah, @ptrsmrn , the target branch for PRs should `master`, not `next`.
I queued the patch, but gh marked this pr as merged too early

ptrsmrn commented:
@scylladb/releng I added backport labels to an already closed PR, will the backport bot still pick them up and open backport PRs or is it too late and I need to open them manually?

yaronkaikov commented:
> @scylladb/releng I added backport labels to an already closed PR, will the backport bot still pick them up and open backport PRs or is it too late and I need to open them manually?

It seems that the base branch of this PR is wrong, i assume this is why the automation is not working , so please do it manually
